MKS WiFi for TFT_LVGL_UI

MKS-Sean opened the pull request:
### Requirements

TFT_LVGL_UI adds the function of MKS WI-FI module

### Description

- Support automatically scan surrounding hotspots and connect on touch screen;
- Support using Cura for wireless file transfer and control, with transfer speed about 100 Kbytes/s
- Support using RepetierHost/PrintRun to control the printer wirelessly;
- Support remote control using MKSCloud APP.
[Please refer to the specific effect video](https://www.facebook.com/Makerbase.mks/videos/660492191332094)

### Benefits

<!-- What does this fix or improve? -->

### Configurations

Add configuration item in Configuration.h file: 
#define USES_MKS_WIFI_FUNCTION

### Related Issues

<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

makerbase-mks committed:
Merge pull request #1 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #2 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #4 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #6 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #7 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #14 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #15 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

MKS-Sean committed:
Merge pull request #16 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

MKS-Sean committed:
Merge pull request #27 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

MKS-Sean committed:
Increase the MKS WI-FI module function of TFT_LVGL_UI

rhapsodyv commented:
Soon I may receive a mks WiFi module, so I can fully test and review it.

thinkyhead committed:
Clean up formatting

thinkyhead committed:
Fix formatting, option name

thinkyhead committed:
also

thinkyhead committed:
Fix typos

thinkyhead commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -142,170 +151,220 @@ void wifi_reset() {
+#include <libmaple/libmaple.h>
+#include <libmaple/gpio.h>
+#include <libmaple/usart.h>
+#include <libmaple/ring_buffer.h>
```
It should be noted that Maple is deprecated and support will be dropped in a future version of PlatformIO.

rhapsodyv replied:
One more point to consider is: the upcoming Nano V3 will be supported only by STM32. So, all this maple code will not work. I'm waiting my MKS Wifi module to test this code and see what we can make more generic.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Misc cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Fix indent

Malderin commented:
@MKS-Sean 
With option #define MKS_WIFI_MODULE - compilation error

```
Marlin\src\lcd\extui\lib\mks_ui\draw_wifi.cpp: In function 'void event_handler(lv_obj_t*, lv_event_t)':
Marlin\src\lcd\extui\lib\mks_ui\draw_wifi.cpp:55:53: error: narrowing conversion of '165' from 'int' to 'char' inside { } [-Wnarrowing]
         char buf[] = { 0xA5, 0x07, 0x00, 0x00, 0xFC };
                                                     ^
Marlin\src\lcd\extui\lib\mks_ui\draw_wifi.cpp:55:53: error: narrowing conversion of '252' from 'int' to 'char' inside { } [-Wnarrowing]
*** [.pioenvs\mks_robin_nano35\src\src\lcd\extui\lib\mks_ui\draw_wifi.cpp.o] Error 1
```

rhapsodyv committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

rhapsodyv committed:
use uint8_t for binary data

rhapsodyv commented:
I tested this PR with a Nano V2 + MKS Wifi Module, and it is working really fine. The file transfer from cura is very fast and I can control my printer directly from cura (I didn't tested all commands). 
The wifi select and connect inside Marlin is really really nice too.

The only issue that should be rewritten in a middle term future is the usage of a hardcoded uart implementation, based maple.

And I would like to add those Wifi options for standard Marlin menus too, of course in a future release.

For me, we can merge it.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_more.cpp:
```diff
@@ -0,0 +1,267 @@
+static void event_handler(lv_obj_t * obj, lv_event_t event) {
+  switch (obj->mks_obj_id) {
+    case ID_CUSTOM_1:
+      if (event == LV_EVENT_CLICKED) {
+      }
+      else if (event == LV_EVENT_RELEASED) {
```
The first `if` isn't needed here if it has no actions. It can be eliminated and just test for `if (event == LV_EVENT_RELEASED)` directly.

Same for the other commands.

rhapsodyv replied:
yes, Scott deleted a lot of this, but they keep popping up over and over again...

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_more.cpp:
```diff
@@ -0,0 +1,267 @@
+      if (event == LV_EVENT_CLICKED) {
+      }
+      else if (event == LV_EVENT_RELEASED) {
+        #if ENABLED(USER_CMD_3_ENABLE)
```
I think the `#if` should surround the entire `case`, to save some code bytes.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_more.cpp:
```diff
@@ -0,0 +1,267 @@
+}
+
+void lv_draw_more(void) {
+  #if ENABLED(USER_CMD_1_ENABLE)
+    lv_obj_t *buttonCustom1;
+  #endif
+  #if ENABLED(USER_CMD_2_ENABLE)
+    lv_obj_t *buttonCustom2;
+  #endif
+  #if ENABLED(USER_CMD_3_ENABLE)
+    lv_obj_t *buttonCustom3;
+  #endif
+  #if ENABLED(USER_CMD_4_ENABLE)
+    lv_obj_t *buttonCustom4;
+  #endif
+  #if ENABLED(USER_CMD_5_ENABLE)
+    lv_obj_t *buttonCustom5;
+  #endif
+  #if ENABLED(USER_CMD_6_ENABLE)
+    lv_obj_t *buttonCustom6;
+  #endif
+  #if ENABLED(USER_CMD_7_ENABLE)
+    lv_obj_t *buttonCustom7;
+  #endif
```
These should all be combined with their initialization. This will save some lines of code and make the file cleaner.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_more.cpp:
```diff
@@ -0,0 +1,267 @@
+  const bool enc_ena = TERN0(HAS_ROTARY_ENCODER, gCfgItems.encoder_enable);
+
+  #if ENABLED(USER_CMD_1_ENABLE)
+    buttonCustom1 = lv_imgbtn_create(scr, "F:/bmp_custom1.bin", INTERVAL_V, titleHeight, event_handler, ID_CUSTOM_1);
```
Is `F:/` a path used on the display? It seems strange to see it here.

rhapsodyv replied:
It's a virtual drive from LVGL. It's used to abstract the load logic. Here, "F:/" is to load data from SPI Flash, and "S:" is to load data from SD.
https://docs.lvgl.io/latest/en/html/overview/file-system.html#add-a-driver

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_cloud_bind.cpp:
```diff
@@ -0,0 +1,255 @@
+  lv_label_set_text(label_ReleaseBind, cloud_menu.unbind);
+  lv_obj_align(label_ReleaseBind, buttonReleaseBind, LV_ALIGN_CENTER, 0, 0);
+
+  //lv_obj_set_event_cb_mks(buttonBack, event_handler,ID_CLOUD_BIND_RETURN, NULL,0);
+  //lv_imgbtn_set_src(buttonBack, LV_BTN_STATE_REL, "F:/bmp_return.bin");
+  //lv_imgbtn_set_src(buttonBack, LV_BTN_STATE_PR, "F:/bmp_return.bin");
+  //lv_imgbtn_set_style(buttonBack, LV_BTN_STATE_PR, &tft_style_label_pre);
+  //lv_imgbtn_set_style(buttonBack, LV_BTN_STATE_REL, &tft_style_label_rel);
```
This file has lots of commented out code. Can it all be deleted?

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_cloud_bind.h:
```diff
@@ -0,0 +1,38 @@
+ */
+#pragma once
+
+#ifdef __cplusplus
+extern "C" { /* C-declarations for C++ */
+#endif
```
Is there a reason for `extern "C"`?

It seems like all the functions are declared and defined in these files, so I expect C++ linkage should be fine.

sjasonsmith replied:
This applies to other headers as well.

rhapsodyv replied:
Some of the functions declared here are used inside the MKS lvl lib (dependency of this code). And the LVGL is pure C code. That's the reason for the `extern "C"`. But we didn't verify what part is really used and part part not, we just let it.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_print_file.cpp:
```diff
@@ -243,7 +243,7 @@ void lv_draw_print_file(void) {
   */
 }
-static char test_public_buf_l[40];
+static char test_public_buf_l[FILE_NUM][SHORT_NEME_LEN + 6];
```
Is `SHORT_NEME` supposed to be `SHORT_NAME`?

sjasonsmith replied:
And why `+ 6`?

Better to used a named constant for a "magic number", so the reason can be seen.

rhapsodyv replied:
to hold SHORT_NEME_LEN + "F:/" ... maybe a few bytes more.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_set.cpp:
```diff
@@ -101,13 +98,8 @@ static void event_handler(lv_obj_t *obj, lv_event_t event) {
-              buf[4] = 0xFC;
-              raw_send_to_wifi(buf, 5);
-
+              uint8_t buf[] = { 0xA5, 0x07, 0x00, 0x00, 0xFC };
```
What is this sequence of bytes? I suggest renaming `buf` to describe the contents.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_ui.cpp:
```diff
@@ -70,8 +70,8 @@ uint8_t printing_rate_update_flag;
-extern uint8_t public_buf[512];
-extern uint8_t bmp_public_buf[17 * 1024];
+extern uint8_t public_buf[513];
+extern uint8_t bmp_public_buf[14 * 1024];
```
What is 14?

thinkyhead replied:
513 and 14 correctly match the non-extern array sizes.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_ui.cpp:
```diff
@@ -137,12 +137,12 @@ void gCfgItems_init() {
-  #if ENABLED(MKS_WIFI_MODULE)
+  gCfgItems.cloud_enable      = false;
+  //#if ENABLED(MKS_WIFI_MODULE)
     gCfgItems.wifi_mode_sel = STA_MODEL;
     gCfgItems.fileSysType   = FILE_SYS_SD;
     gCfgItems.wifi_type     = ESP_WIFI;
-  #endif
+  //#endif
```
Was it a mistake to comment out this `#if`?

rhapsodyv replied:
I think it's a test.. but we can remove the `#if`

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/draw_wifi_list.cpp:
```diff
@@ -93,6 +93,8 @@ void lv_draw_wifi_list(void) {
 
   lv_obj_t *buttonDown = lv_imgbtn_create(scr, "F:/bmp_pageDown.bin", OTHER_BTN_XPIEL * 3 + INTERVAL_V * 4, titleHeight + OTHER_BTN_YPIEL + INTERVAL_H, event_handler, ID_WL_DOWN);
   lv_obj_t *buttonBack = lv_imgbtn_create(scr, "F:/bmp_back.bin", OTHER_BTN_XPIEL * 3 + INTERVAL_V * 4, titleHeight + (OTHER_BTN_YPIEL + INTERVAL_H) * 2, event_handler, ID_WL_RETURN);
+  UNUSED(buttonDown);
+  UNUSED(buttonBack);
```
If these are not used, why save them at all?

It does raise some concern that this might be allocating memory and leaking it. I don't know what lv_imgbtn_create does. It is is returning a pointer into a static pool, then we are fine and you could delete these variables.

If it is calling `new` and returning the result, then we have a problem.

rhapsodyv replied:
lvgl will free all children of a screen, when the is screen be freed. It's have a strong object hierarchy.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/pic_manager.cpp:
```diff
@@ -36,7 +36,7 @@
 
 extern uint16_t DeviceCode;
-extern unsigned char bmp_public_buf[17 * 1024];
+extern unsigned char bmp_public_buf[14 * 1024];
```
More 14s :)

My goal is to be able to read code without having to stop to ask "What is that?". I'm probably getting annoying, but if I have to stop to ask I leave a comment 😁

rhapsodyv replied:
They started with 17 * 1024.... probably with the new code, it run out of run... so they set it to a smaller value. This buffer is the display buffer for LVGL. It's good to be as big as we can set.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/pic_manager.cpp:
```diff
@@ -55,9 +55,10 @@ static const char assets[][LONG_FILENAME_LENGTH] = {
+  // "bmp_zero.bin",
   "bmp_leveling.bin",
   "bmp_filamentchange.bin",
+  "bmp_more.bin",
```
Are these new binaries all present somewhere for users to deploy?

rhapsodyv replied:
Yes, it's automatic. This is the asset load list. There's a script `buildroot/share/PlatformIO/scripts/download_mks_assets.py` , that will download and prepare the assets in a simple folder for the user just copy to the SD.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/printer_operation.cpp:
```diff
@@ -91,10 +93,14 @@ void printer_state_polling() {
-        sprintf_P(public_buf_l, PSTR("G1 Z-%.1f"), gCfgItems.pausePosZ);
-        gcode.process_subcommands_now(public_buf_l);
-        gcode.process_subcommands_now_P(PSTR("G90"));
+        // gcode.process_subcommands_now_P(PSTR("G91"));
+        // ZERO(public_buf_l);
+        // sprintf_P(public_buf_l, PSTR("G1 Z-%.1f"), gCfgItems.pausePosZ);
+        // gcode.process_subcommands_now(public_buf_l);
+        // gcode.process_subcommands_now_P(PSTR("G90"));
```
Can you delete this?

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/tft_Language_en.h:
```diff
@@ -114,6 +114,7 @@
 #define PROBE_Z_SPEED_EN                "Probe Z-axis speed"
 #define ENABLE_EN                       "YES"
 #define DISABLE_EN                      "NO"
+#define LOCKED_EN                       "N/A"
```
Why does this string exist, if it is N/A? I assume this means "Not Applicable"?

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/tft_Language_en.h:
```diff
@@ -275,6 +276,7 @@
 #define PRINT_TEXT_EN           "Printing"
 #define EXTRUDE_TEXT_EN         "Extrusion"
 #define LEVELING_TEXT_EN        "Leveling"
+#define MLEVELING_TEXT_EN       "Leveling"
```
Is there a reason for these to be unique constants, even though they have identical contents?

The same question applies to `TOOL_MLEVELING_EN`.

sjasonsmith replied:
Maybe all four of these should be one constant?

rhapsodyv replied:
M seems to stand for `manual`. They are the same for English, but could not be the same for other lang. Anyway, I think gcc will take care of this for us.

Malderin replied:
#define MLEVELING_TEXT_EN       "Leveling" 
It just needs to be removed. This is the tail from my merge request on the MKS branch. In this code it is not used and will not be used.

Malderin replied:
![11](https://user-images.githubusercontent.com/52313714/101644587-6dd92080-3a46-11eb-8138-6ab73e7ca4e5.jpg)
![22](https://user-images.githubusercontent.com/52313714/101644590-6f0a4d80-3a46-11eb-8c40-d3aa182f03fa.jpg)
all this is not used and should be removed

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp:
```diff
@@ -243,6 +249,34 @@ void my_disp_flush(lv_disp_drv_t * disp, const lv_area_t * area, lv_color_t * co
+      SPI_TFT.tftio.WriteSequence(buf, width);
+
+    W25QXX.init(SPI_QUARTER_SPEED);
+  #else
+    //tft_set_cursor(0, 0);
+    //LCD_setWindowArea((uint16_t)x1, (uint16_t)y1, width, height);
+    //LCD_WriteRAM_Prepare();
+    //#ifdef LCD_USE_DMA_FSMC
+    //  LCD_IO_WriteMultiple(bk_color.full, width * height);
+    //#else
+    //  for (uint32_t i = 0; i < width * height; i++)
+    //    LCD_IO_WriteData(bk_color.full);
+    //#endif
+  #endif
```
Can this be deleted?

sjasonsmith replied:
There is more commented out code in this file at line 422.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifiSerial.cpp:
```diff
@@ -89,7 +89,11 @@ void WifiSerial::begin(uint32 baud, uint8_t config) {
                            config);
   usart_set_baud_rate(this->usart_device, USART_USE_PCLK, baud);
-  usart_enable(this->usart_device);
+  //usart_enable(this->usart_device);
```
delete?

rhapsodyv replied:
I think this is the type of the code that must stay there, even commented... because seems it's not clear why enable is commented...

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -109,8 +115,9 @@ extern IP_PARA ipPara;
+extern char flash_dma_mode;
 
-extern uint8_t bmp_public_buf[17 * 1024];
+extern uint8_t bmp_public_buf[14 * 1024];
```
I have seen this same extern over and over.

It would be better to declare it one time in a header file, then include that header everywhere.
The same applies to other `extern` variables which you use from multiple files.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -142,172 +151,222 @@ void wifi_reset() {
 
 void mount_file_sys(uint8_t disk_type) {
-  if (disk_type == FILE_SYS_SD) {
+  if (disk_type == FILE_SYS_SD)
     TERN_(SDSUPPORT, card.mount());
-  }
   else if (disk_type == FILE_SYS_USB) {
   }
```
You need the curly braces here { }.

`TERN_` evaluated to nothing if SDSUPPORT is false, which would evaluate to this invalid code:

```
if (disk_type == FILE_SYS_SD)
else if (disk_type == FILE_SYS_USB) {
}
```

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -142,172 +151,222 @@ void wifi_reset() {
+  if (flash_dma_mode != dmaMode) {
+    flash_dma_mode = dmaMode;
+    if (flash_dma_mode == 1) {
+      // uint8_t buf[2];
```
Lots of commented out code in this file.

rhapsodyv replied:
This file is new, that probably is the reason :-p

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -142,172 +151,222 @@ void wifi_reset() {
+  while (i) dosName[--i] = '\0';
+  while (*longName) {
+    uint8_t c = *longName++;
+    if (c == '.') {                   // For a dot...
+  if (i == 0) { return false; }
+  else { strcat((char *)dosName,".GCO"); return dosName[0] != '\0'; }
+    }
```
There are formatting issues here that make this hard to read.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -142,172 +151,222 @@ void wifi_reset() {
   }
+}
 
+static bool longName2DosName(const char* longName, uint8_t* dosName) {
+  uint8_t i = 11;
+  while (i) dosName[--i] = '\0';
```
`dosName` is 13 elements long, and you are only clearing 11 of them. This will likely fail to always NUL-terminate strings properly.

rhapsodyv replied:
`dosName` can be 13 or 11. Depends where it can from or is used. For example, these functions work with 11 format `NNNNNNNEEE`:
```
  static bool make83Name(const char* str, uint8_t* name, const char** ptr);
  bool mkdir(SdBaseFile* parent, const uint8_t dname[11]);
  bool open(SdBaseFile* dirFile, const uint8_t dname[11], uint8_t oflag);
```

We need take a look what is correct here.

sjasonsmith commented on Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp:
```diff
@@ -142,172 +151,222 @@ void wifi_reset() {
+      if (c < 0x21 || c == 0x7F) return false;           // Check size, non-printable characters
+      dosName[i++] = (c < 'a' || c > 'z') ? (c) : (c + ('A' - 'a')); // Uppercase required for 8.3 name
+    }
+    if (i >= 5) { strcat((char *)dosName,"~1.GCO"); return dosName[0] != '\0'; }
```
This does not look robust or correct.

What if you have two long filenames which share the same leading 5 characters?

Malderin replied:
The old file will be deleted and replaced with the new one. I tested it.

sjasonsmith reviewed (commented):
I actually thought this code was from @rhapsodyv when I started looking at it. I have made a lot of comments, but they reflect the standard of code that I would want to maintain, and that I would have expected from @rhapsodyv.

I did not realize how long it was, and have not made it through all the files.

My largest concern is the large amount of commented out code throughout these files. I strongly dislike commented out code, and prefer to simply delete it. If these are for future features, they should be kept in another branch until they are ready to enable.

rhapsodyv committed:
minor cleanup

rhapsodyv committed:
save some bytes

rhapsodyv committed:
fix typo

rhapsodyv committed:
more cleanup

thinkyhead committed:
mks-ui cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
buf => cmd_wifi_list

thinkyhead committed:
match types / sizes

thinkyhead commented on Marlin/src/lcd/extui/lib/mks_ui/pic_manager.cpp:
```diff
@@ -373,7 +387,7 @@ uint32_t Pic_Info_Write(uint8_t *P_name, uint32_t P_size) {
 }
 
-uint8_t public_buf[512];
+uint8_t public_buf[513];
```
…and 513 is better than 512.

thinkyhead commented on Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp:
```diff
@@ -79,7 +83,7 @@ extern uint8_t sel_id;
 extern bool flash_preview_begin, default_preview_flg, gcode_preview_over;
 
-uint8_t bmp_public_buf[17 * 1024];
+uint8_t bmp_public_buf[14 * 1024];
```
How do we guarantee that the largest JPG will be under 14 kilobytes? Or 17?

Although we have eschewed the use of `new`, `delete`, `malloc()`, and `free()` in the past, they are okay for short-lived buffers. So, we can allocate file loading buffers as-needed, as long as they are freed as soon as they are used, to prevent heap fragmentation.

rhapsodyv replied:
We don't need handle it. We just tell to lvgl what is the size of this buffer. And it will handle it, and flush the data to the TFT as it need.
The idea of this buffer is: we should use the max we can, that will not interfere with our tasks.

But it can be smaller.

thinkyhead committed:
spacing and spelling

thinkyhead committed:
Consolidate buffers

sjasonsmith commented:
@rhapsodyv & @thinkyhead as I mentioned at the top of my review my contributions to this PR were somewhat accidental, don't feel like you need to wait for me to move it forward.

X-Ryl669 commented:
I'm not a maintainer, so I can't change this PR, but I've fixed some issues with it in my branch and I don't want to pollute with an almost similar PR. Anyway, here are the issues I've spotted so far:

In file `Marlin/src/lcd/extui/lib/mks_ui/draw_baby_stepping.cpp`, line 90 should read:
```
sprintf_P(baby_buf, PSTR("M290 Z%.3f"), -babystep_dist);
```
with the minus sign, else it'll fail to reduce the Z offset.

X-Ryl669 commented:
In draw_print_file.cpp, the function `have_pre_pic` should read:
```
bool have_pre_pic(char *path) {
  #if ENABLED(SDSUPPORT)
    char *ps1, *ps2, *cur_name = strrchr(path, '/');
    if (!cur_name) return false;
    public_buf[511] = 0;
    card.openFileRead(cur_name);
    card.read(public_buf, 511);
    ps1 = strstr((char *)public_buf, ";simage:");
    card.read(public_buf, 511);
    ps2 = strstr((char *)public_buf, ";simage:");
    card.closefile();
    if (ps1 || ps2) return true;
  #endif
Since strstr expect a zero terminated buffer (or it crashes), and it can return null.
  return false;
}
```

X-Ryl669 commented:
Please check this branch for all the fixes: https://github.com/X-Ryl669/MarlinForMKS/tree/MksInclude

X-Ryl669 commented:
Can you merge this so I can submit a PR to fix the spotted errors ? Or maybe one of the maintainer can cherry-pick the fixes from my branch ? 

Waiting while the changes are being pilling up on the bugfix branch will add a lot of work later on to fix conflict

rhapsodyv commented:
> Can you merge this so I can submit a PR to fix the spotted errors ? Or maybe one of the maintainer can cherry-pick the fixes from my branch ?
> 
> Waiting while the changes are being pilling up on the bugfix branch will add a lot of work later on to fix conflict

Hi! Thanks for your fixes. I was out of time this week, but until the week I will test this PR again in my hardware, and incorporate your suggestion/fixes!

thinkyhead committed:
Fix sign on babystep_dist

thinkyhead commented:
> In draw_print_file.cpp, the function have_pre_pic should read:

The buffer is now 513 bytes with the last byte (index 512) guaranteed to be 0.

thinkyhead committed:
Exit if cur_name is null

X-Ryl669 commented:
I'm not fond of 513 bytes buffer, as it's wasting 4 bytes of RAM (since nothing can be stored after the next 32 bit word). What is the interest of using a 513 bytes buffer instead of 512 (and thus limiting read to 511 bytes?) 
The same logic apply, if you initialize the last byte to 0, then there is no need to set it explicitly like I did.

X-Ryl669 commented:
If you correct things by hand, there is also the same issue in function `creat_title_text` in draw_ui.cpp (this times, with the same public buffer but at position 400, meaning that you can't expect the buffer to contain 0 at that position if you either call `have_pre_pic`

MKS-Sean committed:
Merge the revised code of X-Ryl669

MKS-Sean committed:
Restore default configuration

MKS-Sean committed:
Delete the comment code and modify the code format

MKS-Sean commented:
@X-Ryl669 I have merged your code, thank you for your correction.
@sjasonsmith I have deleted the useless comments in the code and corrected some codes, please merge it first, and I will continue to fix the problems you mentioned

X-Ryl669 commented:
Thanks, it looks good to me.

X-Ryl669 committed:
Fix bugs in the file selector screen and icon not being displayed correctly

X-Ryl669 commented:
Can any maintainer merge this code if it's ok ? 
I have issues to report (not related to this PR) and a fix but I don't think this PR is the place to report issues.

MKS-Sean committed:
Resolve conflicting files

MKS-Sean committed:
Resolve conflicting files

MKS-Sean committed:
lvgl ui adds calibration function

MKS-Sean committed:
Resolve file conflicts

MKS-Sean committed:
Merge pull request #29 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #30 from MarlinFirmware/bugfix-2.0.x

Update Bugfix 2.0.x

X-Ryl669 committed:
Adding current position to the move menu (not finished yet)
Add a report of the current position to the move page so it's easier to get information from

X-Ryl669 committed:
Fix missing stack return causing infinite usage of display stack

X-Ryl669 committed:
Fix autoleveling clearing screen when not appropriate
Fix title being cleared when it's too long because the first item in stack does not produce a page text

Rockman18 committed:
[FIX] Case sensitive include on Linux

MKS-Sean committed:
Merge pull request #34 from Rockman18/fix-wifi

[FIX] Case sensitive include on Linux

MKS-Sean committed:
Merge pull request #33 from X-Ryl669/AddStatusInMoveMenu

Add status in move menu

MKS-Sean committed:
Merge pull request #31 from X-Ryl669/FixIcon

Fix bugs in the file selector screen and icon not being displayed correctly

rhapsodyv commented:
@MKS-Sean, I think you should not use this PR to fix arbitrary LVGL issues.

It's starting to become big and unpredictable, and soon (maybe it already does) conflict with another PR.

For example: I already did the fix for gcode preview in #20564, but I'm saying some trying to fix the same issue here, but with a total rewrite of the code, when the issues is just with 4 lines.

This PR will sit here forever if everyday something new appear.

MKS-Sean committed:
Fix CI check problem

MKS-Sean committed:
public_bufpublic_buf

MKS-Sean commented:
@rhapsodyv @X-Ryl669 I restored the code previewed by gcode, it should not conflict now. Code for other functions will not be submitted here.

rhapsodyv commented:
Good. Maybe now is time to do some testing.
@thinkyhead may want take a look in the last changes too.

MKS-Sean committed:
resolve conflict

MKS-Sean committed:
resolve conflicts

MKS-Sean committed:
resolve conflicts

MKS-Sean committed:
Fix wifi transmission problem

MKS-Sean committed:
resolve conflicts

MKS-Sean committed:
resolve conflicts

MKS-Sean commented:
![image](https://user-images.githubusercontent.com/56996910/103335790-211abf80-4ab1-11eb-9841-7a80df10cc1b.png)
![image](https://user-images.githubusercontent.com/56996910/103335807-2e37ae80-4ab1-11eb-9111-e446583aa0dc.png)
![image](https://user-images.githubusercontent.com/56996910/103335822-38f24380-4ab1-11eb-9e81-e89a69a07e93.png)
@rhapsodyv @thinkyhead I passed this PR test, the wifi function is normal, the preview function is normal, it has passed the test

rhapsodyv committed:
file added by mistake?

rhapsodyv committed:
Keep comments for future test env

rhapsodyv committed:
obsolete tft code

Rockman18 commented:
I've tested this PR on Robin Nano v2.0 with PWC. 
If I comment Wifi module, everything seems OK.
`
//#define MKS_WIFI_MODULE  // MKS WiFi module
`

If I activate Wifi module (only uncomment DEFINE), PWC doesn't work anymore. I've to maintain the switch until everything start then  
`
PRINTER HALTED
PLEASE RESET
`

Malderin commented:
Setting/motor-off  causes black screen. 
In file draw_set.cpp in the case ID_S_MOTOR_OFF function lv_clear_set() is superfluous. Correct somebody please.

tpruvot commented:
please check mks_robin env tests... may be the right PR to fix these warnings.... too

X-Ryl669 commented:
@Rockman18 What is your configuration ? Don't you have a conflict with the PWC pin and the wifi's pin ? Can you tell us what pin you have used for the PWC's relay ?

Rockman18 commented:
> @Rockman18 What is your configuration ? Don't you have a conflict with the PWC pin and the wifi's pin ? Can you tell us what pin you have used for the PWC's relay ?

I use nano v2 + MKSPWC + 3dTouch
Here is my config : https://github.com/Rockman18/Marlin/commit/f9a8aa4cd7e1c0c4ce5346e2c542d0194629f10f. 
I use the standard pins_MKS_ROBIN_NANO_V2.h, only uncommented #define MKSPWC.

X-Ryl669 commented:
In the video, it's a Robin Nano v1.2. They are plugging the MKSPWC in the same pin as the BLTouch sensor (that is, Z-Max) so it can't work with both.

I don't understand your schematic here. You can't have both MKSPWC and BLTouch on a Nano v1.2 since the later use the Z-Max pins. In your config, you have redefined the Zmin pin as PC4 but if I understand the schematic, it's supposed to be ZMax pin.
ZMin is PA11
Power detection is PA2
Suicide pin (the one triggering the relay to poweroff the board) is supposed to be PB2 (the lowest connector on the bottom right)
None of them interfere with WIFI pins in anyway. 

So please list the exact connections you've made for all the additional sensors/actuators.

Rockman18 commented:
@X-Ryl669 : At the first time I've added a BLTouch but it doesn't replace the ZMin endstop switch. I used both. So I use ZMax pins for the BLTouch. Now, my Zmin endstop switch is not used anymore and I only use BLTouch but it's still connected to ZMax. 

XMin endstop : PA15
YMin endstop : PA12
ZMin endstop : PA11
BLTouch : PC4 and PA8
Filament Runout : PA4
PWC : PA2 and PB2.

X-Ryl669 commented:
If it's a firmware issue, then the printer must have started to run the firmware. This does not explain why it would not start (whatever the firmware in fact, since it's not running yet).
The "Printer Halter Please Reset" is in 
Marlin/Marlin/src/lcd/extui/lib/mks_ui/draw_error_message.cpp 
in function `lv_draw_error_message` 

If you have a ST-Link V2, you can put a breakpoint on this method and get a backtrace it'll help debugging the issue, you can follow the [instructions from my blog here](https://blog.cyril.by/en/software/debugging-marlin-with-st-link)

If you don't have a ST-Link V2, you'll need to apply my post mortem debugging patches from #20492 and add a crash line in the function (such as `*(int*)0=123;`) so it'll dump the backtrace on the serial link. It'll help diagnose the issue.

rhapsodyv committed:
ID_S_MOTOR_OFF do not change screen

rhapsodyv commented:
> Setting/motor-off causes black screen.
> In file draw_set.cpp in the case ID_S_MOTOR_OFF function lv_clear_set() is superfluous. Correct somebody please.

Fixed

X-Ryl669 commented:
You also need to exclude "ID_CONTINUE" too, see above or https://github.com/makerbase-mks/Marlin-2.x.x/pull/39/files

rhapsodyv committed:
ID_S_CONTINUE and ID_S_MOTOR_OFF

rhapsodyv commented:
For me, `MKSPWC` doesn't work even with WIFI disabled. PA2 is always HIGH for me. Probably because I don't have the proper hardware, I don't know. 

@thinkyhead I think this can be merged, and any other issue can be handled after.

X-Ryl669 commented:
It's logic based on the schematic (the capacitor is loading so you see 3.3V)
![image](https://user-images.githubusercontent.com/3277165/104137626-997d5c80-539e-11eb-9143-46fbea86b40a.png)

What is the schematic of the PWC ?

Rockman18 commented:
@X-Ryl669 : I've found something.... The "Printer Halter Please Reset" is completely normal because it detects the reset button pressed.
In fact, I've to keep the reset button pressed since makerbase splashscreen appears and release it before the main menu appears and everything works.
It's the wifi module update process that is faulty I think, during the try of reading and apply MksWifi.bin.

rhapsodyv commented:
@Rockman18 can you add some logging to try detect _where_ PA2 is changed to high?

Rockman18 commented:
Sorry not really familiar with the code of Marlin yet but I'll try tomorrow. It's 2:00 for me... Sleep time 😉

X-Ryl669 commented:
I think the issue is not when it's pulled high, but when it's pulled down. I'm only *guessing* here, but the schematic with a capacitor means that it's rising up to 3.3V if not pulled down. 
So the only way it can be useful is that the "reset" button connect it to GND when pressed.

In that case, it means that it worked before only by pure luck, the software was able to start faster than the capacitor was charging up so before it triggered the High level, the software switched the PA2 output to use a pulldown probably. Thus the capacitor was charging and then discharging without reaching the level one threshold.

With the new code, the software takes longer to boot and the capacitor triggers the threshold. 

If I'm right, there is nothing to look for in the source code, instead we should move the "change PA2" to pulldown earlier in the boot for the system to kind of "work" like previously.

thinkyhead committed:
Update draw_max_feedrate_settings.cpp

thinkyhead committed:
Reduce

thinkyhead committed:
blerf

X-Ryl669 commented:
Ok, I've found the schematic for the PWC module and it's a bit more complex than what I initially thought:
![image](https://user-images.githubusercontent.com/3277165/104167553-cfe9c480-53fc-11eb-9fde-36748833b6d5.png)
J1 is the actuator (what's labeled PW_OFF on the board, that is not what's plugged on PW_DET), it must show 5V for the power relay to be on (so if it's going to GND, the printers stops)
J2 mirrors the 5V of the PWC module, inverted (so when the PWC module is powered on, there is 0V on the PW_DET pin and 3.3V when it's off, the 3.3V comes from the MKS Nano board)
When you press the switch (labelled S1) you actually force 5V on the relay, triggering the power of the printer.

I'm guessing your "reset" switch is plugged in S2's connector between pin 3 and 4 and when you press and hold the switch, you're driving 5V on R1 and as such, you're pulling PW_DET to ground. If you don't press the reset switch, the R1 value is floating, thus PW_DET is hold to 3.3V due to the capacitor charging on the Nano board (see above).

I guess the code that's testing the value of the PW_DET pin happens later when using this code, and, again, by pure luck with code timing and capacitor's charging time, it happened to work before while it shouldn't have. Now, it's reading a logical one and it's shutting off the printer.

So, to fix this, if I understand correctly, you should put a jumper between pin 3 & 4 of S2 on the PWC board, so that the PW_DET pin is forced to GND whenever 5V is present (that's the purpose of the circuit anyway). 

Please let me know if it fixes the issue for you.

X-Ryl669 commented:
@Rockman18 Did my suggestion worked ?

Rockman18 commented:
> @Rockman18 Did my suggestion worked ?

Sorry I don't have time to test. Perhaps this week-end.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Misc. cleanup

thinkyhead committed:
etc.

thinkyhead committed:
etc.

thinkyhead committed:
final

thinkyhead commented on Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp:
```diff
@@ -123,6 +125,10 @@ void tft_lvgl_init() {
 
+  #if MB(MKS_ROBIN_NANO)
+    OUT_WRITE(PB0, LOW);      // HE1
+  #endif
```
@MKS-Sean — Is this pin prone to being set HIGH on boot? If so, what is the result of that? If this isn't directly associated with the `TFT_LVGL_UI` but more related to the board and its initialization, it may be defined instead as `BOARD_INIT` in the relevant board…
```cpp
#define BOARD_INIT() OUT_WRITE(PB0, LOW)
```
…or…
```cpp
#if HAS_TFT_LVGL_UI
  #define BOARD_INIT OUT_WRITE(PB0, LOW)
#endif
```
…and then it will be initialized early in the `setup()` sequence.
